docs/Datalayer_for_GTM.md

```diff
@@ -0,0 +1,109 @@
+
+This updated approach moves preparing the datalayer out of twig files and provides a way for modules to add or change what is rendered.
+
+## Drupal Structures
+
+Alterations and additions are handled by Drupal's Event system. The final payload is then rendered in twig.
+
+## Altering or Adding to the Datalayer
+
+The `usa_twig_vars` module prepares the initial datalayer array, which sets basic information about the content type, subpage type, node id, and the taxonomy breadcrumb.
+
+Other modules can register an Event listener to change or add items to the datalayer.
+
+First, register the listener in `<MODULE>.services.yml`
+
+```yml
+  MODUILE_datalayer_alter_subscriber:
+    class: '\Drupal\<MODULE>\EventSubscriber\DatalayerAlterSubscriber'
+    tags:
+      - { name: 'event_subscriber' }
+
+```
+
+You'll need a file in your `src` folder in the `EventSubscriber` directory. For the service above, the file name is `DatalayerAlterSubscriber.php` so that Drupal will autoload it.
+
+```php
+<?php
+
+namespace Drupal\<MODULE>\EventSubscriber;
+
+use Drupal\node\Entity\Node;
+use Drupal\usa_twig_vars\Event\DatalayerAlterEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Add benefit category info to datalayer.
+ */
+class DatalayerAlterSubscriber implements EventSubscriberInterface {
+
+  /**
+   * {@inheritDoc}
+   */
+  public static function getSubscribedEvents() {
+    return [
+      DatalayerAlterEvent::EVENT_NAME => 'onDatalayerAlter',
+    ];
+  }
+
+  /**
+   * Adds category information to the datalayer.
+   */
+  public function onDatalayerAlter(DatalayerAlterEvent $event): void {
+    $node = \Drupal::routeMatch()->getParameter('node');
+    $event->datalayer['hasBenefitCategory'] = FALSE;
+
+    // make any changes need to $event->datalayer array
+    $event->datalayer['foo'] = 'bar';
+  }
+
+}
+```
+
+## Twig Templates
+
+The `html.html.twig` template outputs the payload as JSON. There are a number of flags required to render properly, the line should look like the line below. I've wrapped the line here for readability and added comments
+
+```php
+{{ datalayer |
+    // don't render accented characters as unicode entities
+    json_encode(constant('JSON_UNESCAPED_UNICODE')
+        // don't escape slashes in taxonomy URL and elsewhere
+        b-or constant('JSON_UNESCAPED_SLASHES')
+        b-or constant('JSON_PRETTY_PRINT'))
+    | raw }}
+```
+
+## Testing Changes
+
+You can use `scripts/drush/veryify-pagelayer-data.php` to validate that your changes don't change the datalayer output. This is more useful for refactoring the event subscribers or ensuring that additions to the data don't introduce unintended changes. Using the script takes some preparation:
+
+1. *Create Reference Output* The script compares Drupal outputs against a "known-good" data set, the published page CSV file created by a tome export. To prepare this file, you should export your local site using the code from the `dev` branch.
```

Comment:
Please add: "Make sure the URLs in your published_pages.csv point to your development server and not to the default www.usa.gov." 
(I ask for this because I am a slob sometimes and I just take the default settings.)